src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
+      current = this.maxRetries - 1;
+      followRedirect = false;
+    }
+
+    function sleepFor(n) {
+      return new Promise((resolve, reject) => {
+        let delay = Math.pow(2, n) * 100;
+        setTimeout(resolve, delay);
+      });
+    }
+
+    // Note that we've declared current earlier so that we could ensure that
+    // raw streams do not allow for retry
+    for ( ; current < this.maxRetries; current++) {
+      let result = await this.runOnce(opts);
+      if (result.statusCode >= 500 && current !== this.maxRetries - 1) {
+        let errorSummary = `${result.statusCode || '???'}: ${result.statusMessage || '???'}`;
+        debug(`RETRYABLE ERROR ${req.method} ${req.url} --> ${errorSummary}`);
+        await sleepFor(current);
+        continue;
+      } else if (followRedirect) {
+        // https://www.ietf.org/rfc/rfc2616.txt -- 10.3 Redirection 3xx
+        // These are the only 300 series codes that we're sure we can safely
+        // redirect.  Basically, anything which is not a known to be safe
+        // redirect will be returned as the 300 series response rather than
+        // following them.  If a safe to redirect 300 is missing a location
+        // header, then it will be returned as the 300 result as well
+        switch (result.statusCode) {
+          case 301:
+          case 302:
+          case 303:
+            if (req.method === 'GET' || req.method === 'HEAD') {
+              let location = result.headers.location;
+              if (location && currentRedirect < maxRedirects) {
+                return await(this.run({
+                  req: {
+                    url: location,
+                    method: req.method,
+                    headers: req.headers,
+                  },
+                  body,
+                  streamingOutput,
+                  followRedirect,
+                  maxRedirects,
+                  currentRedirect: currentRedirect + 1,
+                }));
+              } else {
+                debug(`WARNING: will only follow ${result.statusCode} with Location header`);
+              }
+            } else {
+              debug(`WARNING: will only follow ${result.statusCode} for GET or HEAD`);
+            }
+            break;
+          default:
+            debug('WARNING: will only follow a 301, 302 or 303 redirect');
```

Comment:
@jhford, so this is why I think we should build the http logic into the client... And not try to split it out like this...

Because the logic for uploads is very specific, and by reducing generality we get to write simpler code that's easier to verify is correct.

Example, not doing redirects is a good thing for uploads...